File: java/MergeTwoSortedLists.java
/*
 * Filename: MergeTwoSortedLists.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Thu 12 Mar 2015 07:52:53 PM EDT
 */

/*
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution {
	public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
		if (l1 == null) {
			return l2;
		}
		
		if (l2 == null) {
			return l1;
		}

		ListNode ret;
		if (l1.val <= l2.val) {
			ret = new ListNode(l1.val);
			ret.next = mergeTwoLists(l1.next, l2);
		} else {
			ret = new ListNode(l2.val);
			ret.next = mergeTwoLists(l1, l2.next);
		}

		return ret;
	}
}

File: cpp/mergeTwoSortedLists.cpp
/*
 * Filename: mergeTwoSortedLists.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Fri 08 Jan 2016 11:22:32 PM EST
 * Description: 
 */

#include <iostream>
#include "listNode.cpp"

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
*/

class Solution {
	public:
		ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
			ListNode* res = new ListNode(0);
			ListNode* dummy = res;

			while (l1 != NULL && l2 != NULL)
			{
				if (l1->val < l2->val)
				{
					res->next = new ListNode(l1->val);
					l1 = l1->next;
				}
				else
				{
					res->next = new ListNode(l2->val);
					l2 = l2->next;
				}
				res = res->next;
			}

			while (l1 != NULL)
			{
				res->next = new ListNode(l1->val);
				res = res->next;
				l1 = l1->next;
			}

			while (l2 != NULL)
			{
				res->next = new ListNode(l2->val);
				res = res->next;
				l2 = l2->next;
			}

			return dummy->next;
		}
};
